Fix "invalid cross-device link" error when running on a different filesystem

lfuelling opened the pull request:
Hi,

this PR fixes #1768 by replacing the call to `os.Rename` in `utils.Move` with a copy followed by a delete. 

I've also used the `utils.Move` function in the thumbnail generator, where `os.Rename` was used directly.

lfuelling committed:
fix cross-device rename errors

Semisol reviewed (changes requested):
I believe this leads to inefficiency if the files are on the same filesystem because there needs to be a full copy of data, and may lead to faster SSD degradation.
Also, this problem could be avoided by making sure the temp file is on the filesystem that is on the same filesystem being moved to.

lfuelling commented:
True, I could probably only do the full copy only when the cross-device link error appears. 

Avoidance of SSD degradation is actually the reason this error appeared for me and why making sure that `/tmp` is on the same filesystem as the application or the data dir is not an option for me.

gabek commented:
Thanks for looking into this.

The one thing I'm concerned about without using rename is the lack of atomic writes.

What would you think about creating a `.tmp` directory inside `data`, so it would be guaranteed to be on the same filesystem, and we use that instead of the system `tmp`? It could be something that gets wiped/recreated on launch.

Or just write things as a temp file in the destination directory and rename it into place, bypassing a separate `tmp` directory completely.

lfuelling commented:
Sure, I thought there was a specific reason for using `os.TempDir`.

gabek commented:
Just convenience and cleanliness, so rogue files aren't being written and it not being clear what they are, and if anything were to stick around, the OS should eventually clean it up. Plus, there's already a temp directory, might as well use it! But I wasn't thinking about this specific limitation.

lfuelling committed:
fallback to copy instead of always using it

lfuelling committed:
use tmp dir in data folder

lfuelling committed:
recreate tmp dir at startup

Semisol commented on utils/utils.go:
```diff
@@ -60,9 +61,40 @@ func Copy(source, destination string) error {
 func Move(source, destination string) error {
-	return os.Rename(source, destination)
+	err := os.Rename(source, destination)
+	if err != nil {
+		log.Warnln("Moving with os.Rename failed, falling back to copy and delete!", err)
+		return moveFallback(source, destination)
+	}
+	return nil
```
Only warn once?

lfuelling replied:
No, `moveFallback(source, destination)` does the copy and delete. If it works there would be no log message.

Semisol replied:
It will spam the log if you warn each time it fails.

lfuelling replied:
Sorry, I don't get what you mean. It has the same logging frequency as my error case in #1768 (one per stream start/stop, two per restart).

Semisol reviewed (changes requested):
Small nitpick.

lfuelling commented:
I've added `./data/tmp` as replacement for `os.TempDir`, except in tests and the updater.
Additionally I added the `os.Rename` back in and made the copy and delete a fallback.

gabek commented:
I think this makes sense, thank you so much @lfuelling!
